tool/tsh/common/scan.go

```diff
@@ -0,0 +1,357 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package common
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"maps"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/alecthomas/kingpin/v2"
+	"github.com/gravitational/trace"
+	"golang.org/x/crypto/ssh"
+
+	accessgraphsecretsv1pb "github.com/gravitational/teleport/api/gen/proto/go/teleport/accessgraph/v1"
+	devicepb "github.com/gravitational/teleport/api/gen/proto/go/teleport/devicetrust/v1"
+	"github.com/gravitational/teleport/api/types/accessgraph"
+	"github.com/gravitational/teleport/lib/accessgraph/secretsservice"
+	"github.com/gravitational/teleport/lib/devicetrust/assert"
+	"github.com/gravitational/teleport/lib/devicetrust/native"
+)
+
+type scanCommands struct {
+	keys *scanKeysCommand
+}
+
+func newScanCommand(app *kingpin.Application) scanCommands {
+	kube := app.Command("scan", "Scan secrets in workforce fleet and report to Teleport")
+	cmds := scanCommands{
+		keys: newScanKeysCommand(kube),
+	}
+	return cmds
+}
+
+type scanKeysCommand struct {
+	*kingpin.CmdClause
+	dirs []string
+	ca   string
+	out  io.Writer
+}
+
+func newScanKeysCommand(parent *kingpin.CmdClause) *scanKeysCommand {
+	c := &scanKeysCommand{CmdClause: parent.Command("keys", "Scan SSH Private Keys in workforce fleet and report to Teleport")}
+	c.Flag("dirs", "Directory to scan for SSH Private Keys").Default("/").StringsVar(&c.dirs)
+	return c
+}
+
+func (c *scanKeysCommand) run(cf *CLIConf) error {
+	ctx := cf.Context
+
+	client, err := secretsservice.NewSecretsScannerServiceClient(
+		ctx,
+		secretsservice.ClientConfig{
+			ProxyServer: cf.Proxy,
+			Insecure:    cf.InsecureSkipVerify,
+			Log:         slog.Default(),
+		})
+
+	deviceCred, err := native.GetDeviceCredential()
+	if err != nil {
+		return trace.Wrap(err, "device not enrolled")
+	}
+
+	fmt.Fprintf(os.Stdout, "Device trust credentials found.\n")
+	fmt.Fprintf(os.Stdout, "Scanning %s.\n", strings.Join(c.dirs, ", "))
+
+	keys := c.scan(ctx, deviceCred.Id)
+	c.print(os.Stdout, keys)
+
+	fmt.Fprintf(os.Stdout, "Creating Teleport client.\n")
+
+	stream, err := client.ReportSecrets(ctx)
+	if err != nil {
+		// FIXME: handle unimplemented
+		return trace.Wrap(err)
+	}
+
+	cremonyProccess, err := assert.NewCeremony()
+	if err != nil {
+		return trace.Wrap(err, "failed to create ceremony")
+	}
+
+	// Run the device authentication ceremony.
+	// If successful, the device will be authenticated and the device can report its secrets.
+	err = cremonyProccess.Run(
+		ctx,
+		reportToAssertstreamAdapter{stream},
+	)
+	if err != nil {
+		return trace.Wrap(err, "failed to run device authentication ceremony")
+	}
+
+	keysSlice := make([]*accessgraphsecretsv1pb.PrivateKey, 0, len(keys))
+	for _, key := range keys {
+		keysSlice = append(keysSlice, key)
+	}
+
+	if err := stream.Send(&accessgraphsecretsv1pb.ReportSecretsRequest{
```

Comment:
In other TAG client streams we usually send in batches. Do we want to do this here? I assume we do not expect to find 1000s of keys, but you never know... 😂